src/Avalonia.Ide.CompletionEngine/XmlParser.cs

```diff
@@ -76,119 +84,231 @@ bool CheckPrev(int caret, string checkFor)
+            if (_elementNameStart >= span.Length)
+            {
+                return "";
+            }
+
+            if (_elementNameEnd != null)
+            {
+                return span.Slice(_elementNameStart, _elementNameEnd.Value - _elementNameStart + 1).ToString();
+            }
+
+            var endTag = _elementNameEnd;
+            for (int i = _elementNameStart; i < span.Length; i++)
+            {
+                char c = span[i];
+                bool isClosingTag = i == _elementNameStart && c == '/';
+                if (!isClosingTag)
+                {
+                    if (char.IsWhiteSpace(c) || c == '/' || c == '>')
+                    {
+                        endTag = i;
+                        break;
+                    }
+                }
+
+                endTag = i + 1;
+            }
+            return span.Slice(_elementNameStart, endTag.Value - _elementNameStart).ToString();
+        }
+
         public static XmlParser Parse(string data)
         {
-            var rv = new XmlParser(data);
-            rv.Parse();
+            return Parse(data.AsMemory());
+        }
+
+        public static XmlParser Parse(ReadOnlyMemory<char> data)
+        {
+            return Parse(data, 0, data.Length);
+        }
+
+        public static XmlParser Parse(ReadOnlyMemory<char> data, int start, int end)
+        {
+            var rv = new XmlParser(data, start);
+            for (var i = start; i < end; i++)
+            {
+                if (!rv.ParseChar())
+                {
+                    break;
+                }
+            }
+
             return rv;
         }
 
+        /// <summary>
+        /// Try parsing until closing tag is found
+        /// </summary>
+        public bool SeekClosingTag()
+        {
+            while (State == ParserState.StartElement)
+            {
+                if (!ParseChar())
+                {
+                    return false;
+                }
+                // leave initial start element
+            }
+
+            while (NestingLevel != 0)
+            {
+                if (!ParseChar())
+                {
+                    return false;
+                }
+                // find next element at same level in document
+            }
+
+            while (State != ParserState.StartElement)
+            {
+                if (!ParseChar())
+                {
+                    return false;
+                }
+                // find start of next element
+            }
+
+            return true;
+        }
+
         public override string ToString()
         {
             return $"State: {State}, TagName: {TagName}, AttributeName: {AttributeName}, Attribute: {AttributeValue}, ContainingTagStart: {ContainingTagStart}";
         }
+
+        /// <summary>
+        /// Clone this parser state
+        /// </summary>
+        /// <returns></returns>
+        public XmlParser Clone()
+        {
+            return (XmlParser)MemberwiseClone();
```

Comment:
This will also "clone" `private Stack<int> _containingTagStart;` by copying a reference.

Reply:
What you think is the best way to deal with this problem? Assign all fields by hand, or keep MemberwiseClone and copy stack by hand?

Reply:
I guess the second one is fine.

Reply:
Fixed.

Btw. nice catch with this.